src/components/select/select.tsx

```diff
@@ -325,43 +386,128 @@ export function SelectOptions( {
 
 	// Render children based on the search keyword.
 	const renderChildren = useMemo( () => {
-		return Children.map( children, ( child, index ) => {
+		// Track actual visible groups after filtering
+		let visibleGroups = 0;
+		let totalGroups = 0;
+
+		// First pass - count total visible groups
+		Children.forEach( children, ( child ) => {
+			if ( isValidElement( child ) && child.type === SelectOptionGroup ) {
+				const hasVisibleChildren = Children.toArray(
+					child.props.children
+				).some( ( groupChild ) => {
+					if ( ! isValidElement( groupChild ) ) {
+						return false;
+					}
+
+					if ( searchKeyword ) {
+						const valueProp = groupChild.props.value;
+						if ( typeof valueProp === 'object' ) {
+							return valueProp[ searchBy ]
+								.toLowerCase()
+								.includes( searchKeyword.toLowerCase() );
+						}
+						return valueProp
+							.toLowerCase()
+							.includes( searchKeyword.toLowerCase() );
```

Comment:
we can add a common utility function here.

Reply:
Both conditions are different and used for different purposes.